diff --git a/lib/python/Plugins/SystemPlugins/AnimationSetup/plugin.py b/lib/python/Plugins/SystemPlugins/AnimationSetup/plugin.py
index 80327d2..7a2914c 100644
--- a/lib/python/Plugins/SystemPlugins/AnimationSetup/plugin.py
+++ b/lib/python/Plugins/SystemPlugins/AnimationSetup/plugin.py
@@ -12,7 +12,7 @@ from enigma import setAnimation_current, setAnimation_speed, setAnimation_curren
 # default = slide to left
 g_default = {
 	"current": 6,
-	"speed"  : 20,
+	"speed"  : 25,
 	"listbox": "0",
 }
 g_max_speed = 30
