diff --git a/lib/python/Components/Harddisk.py b/lib/python/Components/Harddisk.py
index c455e5f..f88b857 100755
--- a/lib/python/Components/Harddisk.py
+++ b/lib/python/Components/Harddisk.py
@@ -78,7 +78,8 @@ def isUsbStorage(filename):
 		phyPath = os.path.realpath('/sys/block/' + dev[:3])
 		for x in glob.glob("/sys/bus/usb/devices/usb*"):
 			if phyPath.find(os.path.realpath(x)) != -1:
-				return True
+				if phyPath.find("f0b00500.ehci_v2/usb2/2-1") == -1:
+					return True
 
 	return False
 
